### PR TITLE
feat: [CDS-78464]: API for post creation on Canny

### PR DESCRIPTION
**Pull request by ashinsabu3**
## Describe your changes

API under NG Manager to create a post on canny. This will be used to redirect tickets of type "FEATURE_REQUEST" to canny instead of Zendesk.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53358)

### CODE REVIEW

**Comment by ashinsabu3** [comment]
trigger uts

**Comment by ashinsabu3** [comment]
trigger ss

**Comment by ashinsabu3** [comment]
trigger uts

File: 120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java [hunk 1]
```diff
@@ -32,13 +35,27 @@
 @OwnedBy(HarnessTeam.CDP)
 public class CannyClient {
   private CannyConfig cannyConfig;
+
   public static final String CANNY_BASE_URL = "https://canny.io";
```
[the remark this answers is not in the document]
Comment (ashinsabu3) [reply 120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java:39]: Sure, this seems like a good idea for making it cleaner

[the remark this answers is not in the document]
Comment (ashinsabu3) [reply 120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java:39]: I'm thinking the BASE_URL and API_BASE_URL can be kept in the same class

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 58 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnyBER7M6mgoseHI7D:-NfnyBER7M6mgoseHI7E:b-dqopad) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L58)):*
> ```Java
> 
>   @VisibleForTesting
>   public OkHttpClient okHttpClient = new OkHttpClient.Builder().retryOnConnectionFailure(true).build();
> ```

this need not be visible for testing
- its already public
- this does not have a business logic to test

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 81 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnyRyf33daEdcHIi4q:-NfnyRyf33daEdcHIi4r:b-ud6t3q) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L81)):*
> ```Java
>         String name = boardNode.get(NAME_NODE).asText();
> 
>         // ------------------------------------------------------------------------------------------------
> ```

nit: these line markers are not needed

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 88 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnyOstAwazge3cGd-4:-NfnyOsuA0EHOdjt95zI:b-ud6t3q) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L88)):*
> ```Java
>           continue;
>         }
>         // ------------------------------------------------------------------------------------------------
> ```

nit: these line markers are not needed

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 108 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnymH12Azcf2FJ0lTb:-NfnymH12Azcf2FJ0lTc:bjsxteg) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L108)):*
> ```Java
>       Response response = createCannyPost(authorId, boardId, title, details);
>       if (!response.isSuccessful()) {
>         log.error("Request to canny failed trying to create post. Response body: {}", response.body().string());
> ```

does this contain the status code from canny ?
Also, we should not send the response as is, to out users who are creating tickets
We should try to parse the response body

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 119 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-Nfnz-5X8GtiD8itjSDT:-Nfnz-5X8GtiD8itjSDU:bnj7hl0) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L119)):*
> ```Java
>             postDetailsResponse.body().string());
>         throw new UnexpectedException("Request to canny failed trying to retrieve post details. Response body: "
>             + postDetailsResponse.body().string());
> ```

same as above

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 137 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzGfM1nNp9lIJ9wyY:-NfnzGfM1nNp9lIJ9wyZ:bz03c2v) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L137)):*
> ```Java
>       // if user doesn't exist on canny(checked through email since it is a unique entity on harness), create user
>       if (!response.isSuccessful() && response.code() != 500
>           && jsonResponse.get(ERROR_NODE).asText().equals(INVALID_EMAIL_ERROR)) {
> ```

constant should be on the left side of a comparator
INVALID_EMAIL_ERROR.equals(jsonResponse.get(ERROR_NODE).asText())

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 136 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzDjFAh_qnPlheYbL:-NfnzDjFAh_qnPlheYbM:br7okn4) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L136)):*
> ```Java
> 
>       // if user doesn't exist on canny(checked through email since it is a unique entity on harness), create user
>       if (!response.isSuccessful() && response.code() != 500
> ```

why do we need this condition `response.code() != 500` ?

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 145 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzTPb5pSlLVpCPuqk:-NfnzTPb5pSlLVpCPuql:b2cpufo) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L145)):*
> ```Java
>               response.body().string());
>           throw new UnexpectedException("Request to canny failed trying to create user during getPostCreationAuthorId."
>               + response.body().string());
> ```

we should not send the response as is, to out users who are creating tickets  
We should try to parse the response body

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 175 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzWEfED7wMrXfSPHV:-NfnzWEfED7wMrXfSPHW:b-bluwmi) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L175)):*
> ```Java
>       throw new UnexpectedException("Exception occurred while making call to canny to fetch boards: ", e);
>     }
>   }
> ```

new line for readability

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 194 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzdIN9n8XGhBkyRj4:-NfnzdIN9n8XGhBkyRj5:bhv5zlq) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L194)):*
> ```Java
>                                             .add("userID", userId)
>                                             .add("email", email)
>                                             .add("name", name);
> ```

all these fields should be declared as String constants

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 211 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzjqI3iu40cewPPfS:-NfnzjqI3iu40cewPPfT:bhv5zlq) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L211)):*
> ```Java
>                                              .add("boardID", boardId)
>                                              .add("details", details)
>                                              .add("title", title);
> ```

all these fields should be declared as String constants

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 39 at r1](https://reviewable.io/reviews/harness/harness-core/53358#gh-1340269440:-Nfnzunj1oZiCfhqYcfb:b-fzvgo2) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L39)):*
<details><summary><i>Previously, ashinsabu3 (Ashin Sabu) wrote…</i></summary><blockquote>

I'm thinking the BASE_URL and API_BASE_URL can be kept in the same class
</blockquote></details>

We can move all of them to a new class, should be fine

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 3 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53358)* status: Some of the checks are too old. 4 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained.

**Comment by ashinsabu3** [comment]
> _[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 108 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnymH12Azcf2FJ0lTb:-NfnymH12Azcf2FJ0lTc:bjsxteg) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L108)):_
> 
> > ```java
> >       Response response = createCannyPost(authorId, boardId, title, details);
> >       if (!response.isSuccessful()) {
> >         log.error("Request to canny failed trying to create post. Response body: {}", response.body().string());
> > ```
> 
> does this contain the status code from canny ? Also, we should not send the response as is, to out users who are creating tickets We should try to parse the response body

This response only contains a single message "error":"<error why given operation couldn't complete"> in the body, so I assumed parsing separately would add unnecessary extra step. It never exposes any other information other than the error.

**Comment by ashinsabu3** [comment]
> _[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 136 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzDjFAh_qnPlheYbL:-NfnzDjFAh_qnPlheYbM:br7okn4) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L136)):_
> 
> > ```java
> >       // if user doesn't exist on canny(checked through email since it is a unique entity on harness), create user
> >       if (!response.isSuccessful() && response.code() != 500
> > ```
> 
> why do we need this condition `response.code() != 500` ?

we must create a new user only if the exact response code matches a 400 with a user not found message in the body, in case there is some issue on the canny side for this response which returns a 500 internal error or something of the sort, we must not attempt to create a user because user may already exist but canny's api failed on their side due to some reason

**Review by ashinsabu3 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53358)* status: Some of the checks are too old. 4 file(s) to review. 6. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 137 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzGfM1nNp9lIJ9wyY:-NfoBH8Q7brHmoraFZaF:bs7wumb) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L137)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

constant should be on the left side of a comparator
INVALID_EMAIL_ERROR.equals(jsonResponse.get(ERROR_NODE).asText())
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 175 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzWEfED7wMrXfSPHV:-NfoBMBF2TLOhRFM3HYI:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L175)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

new line for readability
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 194 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzdIN9n8XGhBkyRj4:-NfoBMrO0W0fwG7n0QRj:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L194)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

all these fields should be declared as String constants
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 211 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzjqI3iu40cewPPfS:-NfoBNQB5R1tPgw2FDEd:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L211)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

all these fields should be declared as String constants
</blockquote></details>

Done.

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 108 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnymH12Azcf2FJ0lTb:-NfoNFGA2nkX6mic6S1u:b96vsi1) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L108)):*
<details><summary><i>Previously, ashinsabu3 (Ashin Sabu) wrote…</i></summary><blockquote>

This response only contains a single message "error":"<error why given operation couldn't complete"> in the body, so I assumed parsing separately would add unnecessary extra step. It never exposes any other information other than the error.
</blockquote></details>

could you please provide a screenshot of this looks on our UI if it is possible to induce an error ?

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 136 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzDjFAh_qnPlheYbL:-NfoOj1VE4NL2HyW_pWN:b-gli26e) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L136)):*
<details><summary><i>Previously, ashinsabu3 (Ashin Sabu) wrote…</i></summary><blockquote>

we must create a new user only if the exact response code matches a 400 with a user not found message in the body, in case there is some issue on the canny side for this response which returns a 500 internal error or something of the sort, we must not attempt to create a user because user may already exist but canny's api failed on their side due to some reason
</blockquote></details>

so we should check with code == 400 right ?

**Comment by ashinsabu3** [comment]
> _[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 108 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnymH12Azcf2FJ0lTb:-NfoNFGA2nkX6mic6S1u:b96vsi1) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L108)):_
> 
> Previously, ashinsabu3 (Ashin Sabu) wrote…
> > This response only contains a single message "error":"<error why given operation couldn't complete"> in the body, so I assumed parsing separately would add unnecessary extra step. It never exposes any other information other than the error.
> 
> could you please provide a screenshot of this looks on our UI if it is possible to induce an error ?

Yup, just changed (!response.isSuccesful) to (response.isSuccessful)
image
response message only shows the finally thrown exception after all catching is done, at catch and rethrow
https://github.com/harness/harness-core/pull/53358/files#diff-0ab10ee938d063bcd00a3e60c24aec75f906aa0d11ff33280426774978c4ed19R104-R108

**Comment by ashinsabu3** [comment]
> _[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 136 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzDjFAh_qnPlheYbL:-NfoOj1VE4NL2HyW_pWN:b-gli26e) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L136)):_
> 
> Previously, ashinsabu3 (Ashin Sabu) wrote…
> so we should check with code == 400 right ?

Ack, changing this.

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 119 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-Nfnz-5X8GtiD8itjSDT:-NfyDcPo4KtomH3cj7yF:bci7tnc) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L119)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

what happens if canny changes error response and jsonResponse.get(CannyClientConstants.ERROR_NODE) is null 
We should check if the error node is present, then use it
or else fallback to sending the complete node

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 136 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzDjFAh_qnPlheYbL:-NfyDqb52tzEHzdvYUi-:bt60l0q) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L136)):*
<details><summary><i>Previously, ashinsabu3 (Ashin Sabu) wrote…</i></summary><blockquote>

Ack, changing this.
</blockquote></details>

Better would be to check if the status is in the range 400-499 because these http codes are reserved for user errors

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 2 files at r2, 1 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53358)* status: 5 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained.

**Comment by ashinsabu3** [comment]
> _[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 136 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzDjFAh_qnPlheYbL:-NfyDqb52tzEHzdvYUi-:bt60l0q) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L136)):_
> 
> Previously, ashinsabu3 (Ashin Sabu) wrote…
> > Ack, changing this.
> 
> Better would be to check if the status is in the range 400-499 because these http codes are reserved for user errors

I disagree here, canny only sends 400 for user not found, there are other codes like 429(rate limit), 408(timeout), where we must not attempt to create a new user

**Review by ashinsabu3 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53358)* status: 5 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 39 at r1](https://reviewable.io/reviews/harness/harness-core/53358#gh-1340269440:-NfoHjaj42bOrE5pMoNz:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L39)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

We can move all of them to a new class, should be fine
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 119 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-Nfnz-5X8GtiD8itjSDT:-NfypqfI2htmkG3cO4cX:bqr77bl) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L119)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

what happens if canny changes error response and jsonResponse.get(CannyClientConstants.ERROR_NODE) is null 
We should check if the error node is present, then use it
or else fallback to sending the complete node
</blockquote></details>

Added a check for this
___
*[`120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java` line 145 at r1](https://reviewable.io/reviews/harness/harness-core/53358#-NfnzTPb5pSlLVpCPuqk:-Nfypw0H6HC3mItBtYxc:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/4558cae4219bace617c5ade7e1c2bbc9874872fe/120-ng-manager/src/main/java/io/harness/ng/support/client/CannyClient.java#L145)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

we should not send the response as is, to out users who are creating tickets  
We should try to parse the response body
</blockquote></details>

Done.

**Review by yogesh-chauhan — DISMISSED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53358)* status: 5 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Comment by ashinsabu3** [comment]
trigger uts, review

**Comment by ashinsabu3** [comment]
trigger uts

**Comment by ashinsabu3** [comment]
trigger uts

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53358)* status: 5 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by mankrit-singh — APPROVED** [review]

**Comment by ashinsabu3** [comment]
trigger review

**Pull request merged**